addons/controllers/packageinstallstatus_controller.go

```diff
@@ -0,0 +1,441 @@
+		Reason:             pkgiCondition.Reason,
+		LastTransitionTime: metav1.NewTime(time.Now().UTC().Truncate(time.Second)),
+	}
+
+	// only add a new condition entry for the PackageInstall in the clusterBootstrapStatus in case it doesn't already exist.
+	// If it does, just update it with the new condition.
+	// Note that we did not simply use cluster API's util function conditions.Set(clusterBootstrap, &condition) cause our condition types are generated by prefixing Carvel's condition types with pkgi name and we need
+	// to only consider condition types' prefix (pkgi name) rather than the full condition type for condition's equality check and custom comparison logic is net implemented in CAPI's condition util Set() as of now
+	var conditionExists bool
+	for i, existingCond := range clusterBootstrap.Status.Conditions {
+		if !strings.Contains(string(existingCond.Type), strings.Title(pkgShortname)) {
+			continue
+		}
+		conditionExists = true
+		if !util.HasSameState(&clusterBootstrap.Status.Conditions[i], &condition) {
+			clusterBootstrap.Status.Conditions[i] = condition
+		}
+	}
+	if !conditionExists {
+		clusterBootstrap.Status.Conditions = append(clusterBootstrap.Status.Conditions, condition)
+	}
+
+	return nil
+}
+
+// removeConditionIfExistsForPkgName removes the corresponding condition for the provided pkgRefName from the clusterBootstrapStatus if existing
+func (r *PackageInstallStatusReconciler) removeConditionIfExistsForPkgName(clusterBootstrap *runtanzuv1alpha3.ClusterBootstrap, pkgRefName string) {
+	for i, existingCond := range clusterBootstrap.Status.Conditions {
+		pkgShortname := strings.Split(pkgRefName, ".")[0]
+		if strings.Contains(string(existingCond.Type), strings.Title(pkgShortname)) {
+			clusterBootstrap.Status.Conditions = append(clusterBootstrap.Status.Conditions[:i], clusterBootstrap.Status.Conditions[i+1:]...)
+		}
+	}
+}
+
+// watchPackageInstalls sets a remote watch on the provided cluster on the Kind resource
+func (r *PackageInstallStatusReconciler) watchPackageInstalls(cluster *clusterapiv1beta1.Cluster, log logr.Logger) error {
+	// If there is no tracker, don't watch remote package installs
+	if r.tracker == nil {
+		return nil
+	}
+
+	return r.tracker.Watch(r.ctx, remote.WatchInput{
+		Name:         "watchPackageInstallStatus",
+		Cluster:      clusterapiutil.ObjectKey(cluster),
+		Watcher:      r.controller,
+		Kind:         &kapppkgiv1alpha1.PackageInstall{},
+		EventHandler: handler.EnqueueRequestsFromMapFunc(r.pkgiToCluster),
+		Predicates:   []predicate.Predicate{r.pkgiIsManagedAndStatusChanged(log)},
+	})
+}
+
+// pkgiIsManagedAndStatusChanged returns a predicate.Predicate that filters pkgi objects which their status has changed
+func (r *PackageInstallStatusReconciler) pkgiIsManagedAndStatusChanged(log logr.Logger) predicate.Funcs {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return r.processPkgiStatus(e.Object, log.WithValues("predicate", "createEvent"))
```

Comment:
You might want to rename this func to, like the predicate maybe `processPkgiIsManagedAndStatusChanged`